docs/architecture/state.md

```diff
@@ -1,91 +1,111 @@
+Private notes greatly reduce storage requirements and thus result in lower fees. They add only $64$ bytes to the state ($32$ bytes when produced and $32$ bytes when consumed).
+
+At high throughput (e.g., 1K TPS), the note database could grow by about 1TB/year. However, only unconsumed public notes and enough information to construct membership proofs must be stored explicitly. Private notes, as well as consumed public notes, can be discarded. This solves the issue of infinitely growing note databases.
 
 ![Architecture core concepts](../img/architecture/state/note-db.png)
 
-As described in [the notes section](notes.md), there are two types of notes:
+### Nullifiers database
+
+Nullifiers map one-to-one to existing notes, tracking whether a note has been consumed.
 
-* Public notes where the entire note content is recorded in the state.
-* Private notes where only a note's hash is recorded in the state.
+To prove that a note has not been consumed, the operator must provide a Merkle path to the corresponding node and show that the node’s value is `0`. Since nullifiers are $32$ bytes each, the Sparse Merkle Tree height must be sufficient to represent all possible nullifiers. Operators must maintain the entire nullifier set to compute the new tree root after inserting new nullifiers.
 
-As with accounts, there is a strong incentive to use private notes as they result in lower fees. This is also beneficial to the network as a private note adds only $64$ bytes to the state ($32$ bytes when it is produced, and $32$ bytes when it is consumed).
+> **Note**
+> - Nullifiers in Miden break linkability between privately stored notes and their consumption details. To know the [note’s nullifier](notes.md#note-nullifier-ensuring-private-consumption), one must know the note’s data.
 
-Using a Merkle Mountain Range (append-only accumulator) is important for two reasons:
+![Architecture core concepts](../img/architecture/state/nullifier-db.png)
 
-1. Membership witnesses (that a note exists in the database) against such an accumulator needs to be updated very infrequently.
-2. Old membership witnesses can be extended to a new accumulator value, but this extension does not need to be done by the original witness holder.
+## Additional information
 
-Both of these properties are needed for supporting local transactions using client-side proofs and privacy. In an append-only data structure, witness data does not become stale when the data structure is updated. That means users can generate valid proofs even if they don’t have the latest state of this database; so there is no need to query the operator on a constantly changing state.
+### Public shared state
```

Comment:
This is super important. A lot of ppl ask about it

Reply:
What is super important?